src/Components/Components/test/Routing/QueryParameterValueSupplierTest.cs

```diff
@@ -0,0 +1,503 @@
+            [Parameter, SupplyParameterFromQuery(Name = "b")] public DateTime ValueAsDateTime { get; set; }
+            [Parameter, SupplyParameterFromQuery(Name = "A")] public long ValueAsLong { get; set; }
+        }
+
+        [Fact]
+        public void CannotMapSingleQueryParameterToMultipleProperties()
+        {
+            var ex = Assert.Throws<InvalidOperationException>(
+                () => QueryParameterValueSupplier.ForType(typeof(MapSingleQueryParameterToMultipleProperties)));
+            Assert.Contains("declares more than one mapping for the query parameter 'a'.", ex.Message, StringComparison.OrdinalIgnoreCase);
+        }
+
+        class UnsupportedType : ComponentBase
+        {
+            [Parameter, SupplyParameterFromQuery] public int IntValid { get; set; }
+            [Parameter, SupplyParameterFromQuery] public object ObjectValue { get; set; }
+        }
+
+        [Fact]
+        public void RejectsUnsupportedType()
+        {
+            var ex = Assert.Throws<NotSupportedException>(
+                () => QueryParameterValueSupplier.ForType(typeof(UnsupportedType)));
+            Assert.Equal("Querystring values cannot be parsed as type 'System.Object'.", ex.Message);
+        }
+
+        [Theory]
+        [InlineData(nameof(ValidTypes.BoolVal), "abc", typeof(bool))]
+        [InlineData(nameof(ValidTypes.DateTimeVal), "2020-02-31", typeof(DateTime))]
+        [InlineData(nameof(ValidTypes.DecimalVal), "1.2.3", typeof(decimal))]
+        [InlineData(nameof(ValidTypes.DoubleVal), "1x", typeof(double))]
+        [InlineData(nameof(ValidTypes.FloatVal), "1e1000", typeof(float))]
+        [InlineData(nameof(ValidTypes.GuidVal), "123456-789-0", typeof(Guid))]
+        [InlineData(nameof(ValidTypes.IntVal), "5000000000", typeof(int))]
+        [InlineData(nameof(ValidTypes.LongVal), "this+is+a+long+value", typeof(long))]
+        [InlineData(nameof(ValidTypes.NullableBoolVal), "abc", typeof(bool?))]
+        [InlineData(nameof(ValidTypes.NullableDateTimeVal), "2020-02-31", typeof(DateTime?))]
+        [InlineData(nameof(ValidTypes.NullableDecimalVal), "1.2.3", typeof(decimal?))]
+        [InlineData(nameof(ValidTypes.NullableDoubleVal), "1x", typeof(double?))]
+        [InlineData(nameof(ValidTypes.NullableFloatVal), "1e1000", typeof(float?))]
+        [InlineData(nameof(ValidTypes.NullableGuidVal), "123456-789-0", typeof(Guid?))]
+        [InlineData(nameof(ValidTypes.NullableIntVal), "5000000000", typeof(int?))]
+        [InlineData(nameof(ValidTypes.NullableLongVal), "this+is+a+long+value", typeof(long?))]
+        public void RejectsUnparseableValues(string key, string value, Type targetType)
+        {
+            var ex = Assert.Throws<InvalidOperationException>(
+                () => GetSuppliedParameters<ValidTypes>($"?{key}={value}"));
+            Assert.Equal($"Cannot parse the value '{value.Replace('+', ' ')}' as type '{targetType}' for '{key}'.", ex.Message);
```

Comment:
Wouldn't something like this mean that I can type something invalid on the URL and cause the circuit/app to crash?

Reply:
Should we instead consider setting a default value instead of throwing? (Modelbinding does this).

Reply:
You can indeed cause the circuit (not app) to be terminated by supplying nonparseable data. I did consider setting a default value but that makes it harder for the app developer to notice and understand if they are setting wrongly-formatted parameters (e.g., unsupported date formats).

Since an application would never have reason to provide invalid parameter values to itself, this should only throw if someone is deliberately modifying the URL directly, and so I considered ending the circuit to be a reasonable outcome. This doesn't let someone malicious do anything new, as they could always kill their own circuit with an unhandled exception, e.g., by making a malformed JS interop call.

Reply:
That's fair, the only thing that I worry about a bit, is that we don't give the developer any hook to avoid terminating the app, and I worry a bit that you can run into this situation by mistake quite often (like for example when binding the query string to a textbox or similar).

If you've considered those things and are ok, then it's fine. Worst case scenario I suspect you can use the Error boundaries component to prevent the circuit from completely crashing. We can wait until we receive feedback to determine if this is a problem or not, and offer some sort of switch/knob/error handler to allow you to deal with errors in whatever way you find best.